packages/vite-plugin-rsc/src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+import { Plugin, ResolvedConfig, Manifest } from "vite";
+import path from "node:path";
+import { pathToFileURL } from "node:url";
+import { existsSync, readFileSync } from "node:fs";
+interface ASTNode {
+  type: string;
+  start: number;
+  end: number;
+  body?: Array<ASTNode>;
+  id?: ASTNode;
+  expression?: ASTNode;
+  declaration?: ASTNode;
+  declarations?: Array<ASTNode>;
+  name?: string;
+  specifiers?: Array<ASTNode>;
+
+  value?: string;
+  exported?: ASTNode;
+}
+
+/**
+ * Checks if the node has a string literal at the top level that matches the statement
+ */
+const hasPragma = (ast: ASTNode, statement: string) =>
+  ast.body?.some((node) => {
+    return (
+      node.type === "ExpressionStatement" &&
+      node.expression?.type === "Literal" &&
+      node.expression.value === statement
+    );
+  });
+
+/**
+ * Finds all the named and default exports of a module
+ */
+
+const getExports = (ast: ASTNode) => {
+  const exports: Array<string> = [];
+  ast.body?.forEach((node) => {
+    if (node.type === "ExportDefaultDeclaration") {
+      exports.push("default");
+    }
+    if (node.type === "ExportNamedDeclaration") {
+      if (node.declaration?.type === "VariableDeclaration") {
+        node.declaration?.declarations?.forEach((declaration) => {
+          const name = declaration?.id?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+        return;
+      }
+
+      if (node.declaration?.type === "FunctionDeclaration") {
+        const name = node.declaration?.id?.name;
+        if (name) {
+          exports.push(name);
+        }
+        return;
+      }
+
+      if (node.specifiers?.length) {
+        node.specifiers.forEach((specifier) => {
+          const name = specifier?.exported?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+      }
+    }
+  });
+  return exports;
+};
+
+export default function plugin({
+  serverDist = "dist/server",
+  clientDist = "dist/static",
+}: {
+  serverDist?: string;
+  clientDist?: string;
+}): Plugin {
```

Comment:
These are not read from the config because we need to know the value of the server dist during client build and vice versa. This does _not_ set the value for the builds - it's just for reference so we can find the manifests.